File: backend/app/application/mobile/journey_destination.py
"""Public place projection used by the trip globe; never an inferred airport."""

from typing import Literal, Self

from pydantic import BaseModel, Field, model_validator


class JourneyDestination(BaseModel):
    label: str = Field(min_length=1, max_length=255)
    country: str = Field(min_length=1, max_length=160)
    country_code: str = Field(pattern=r"^[A-Z]{2}$")
    latitude: float = Field(ge=-90, le=90, allow_inf_nan=False)
    longitude: float = Field(ge=-180, le=180, allow_inf_nan=False)
    place_type: Literal["country", "state", "city"]


class MobileJourneyDestinationResponse(BaseModel):
    status: Literal["resolved", "not_found", "ambiguous", "unavailable", "not_configured"]
    destination: JourneyDestination | None = None
    attribution: Literal["© OpenStreetMap contributors"] = "© OpenStreetMap contributors"
    attribution_url: Literal["https://www.openstreetmap.org/copyright"] = (
        "https://www.openstreetmap.org/copyright"
    )
    retry_after_seconds: int | None = Field(default=None, ge=1, le=86_400)

    @model_validator(mode="after")
    def validate_resolution(self) -> Self:
        if (self.status == "resolved") != (self.destination is not None):
            raise ValueError("Coordinates require a resolved destination")
        return self

File: mobile/src/features/journey-globe/model/journey-destination.ts
import { z } from 'zod';

import { routeToDestination } from './journey-route';

export const JourneyDestinationSchema = z.object({
  status: z.enum(['resolved', 'unavailable', 'ambiguous', 'not_found', 'not_configured']),
  destination: z.object({
    label: z.string().min(1).max(255),
    country: z.string().min(1).max(160),
    country_code: z.string().min(2).max(3),
    latitude: z.number().finite().min(-90).max(90),
    longitude: z.number().finite().min(-180).max(180),
    place_type: z.enum(['city', 'state', 'country', 'region']),
  }).nullable(),
  attribution: z.string().max(150),
  attribution_url: z.literal('https://www.openstreetmap.org/copyright'),
  retry_after_seconds: z.number().nonnegative().nullable().optional(),
});
export type JourneyDestination = z.infer<typeof JourneyDestinationSchema>;
export type JourneyLookupStatus = JourneyDestination['status'] | 'resolving';

export function routeFromLookup(tripId: string, lookup: JourneyDestination) {
  if (lookup.status !== 'resolved' || !lookup.destination) return null;
  const place = lookup.destination;
  return routeToDestination(tripId, {
    city: place.label, country: place.country, countryCode: place.country_code.toUpperCase(),
    latitude: place.latitude, longitude: place.longitude, placeType: place.place_type,
  }, lookup.attribution);
}
